example-async.js

```diff
@@ -0,0 +1,158 @@
+'use strict'
+
+const util = require('util')
+const Fastify = require('fastify')
+
+function build (opts) {
+  const fastify = Fastify(opts)
+
+  fastify
+    .register(require('fastify-jwt'), { secret: 'supersecret' })
+    .register(require('fastify-leveldb'), { name: 'authdb-async' })
+    .register(require('./fastify-auth'))
+    .after(routes)
+
+  fastify.decorate('verifyJWTandLevel', verifyJWTandLevel)
+  fastify.decorate('verifyUserAndPassword', verifyUserAndPassword)
+
+  function verifyJWTandLevel (request, reply, done) {
+    const jwt = this.jwt
+    const level = this.level
+
+    if (request.body && request.body.failureWithReply) {
+      reply.code(401).send({ error: 'Unauthorized' })
+      return done(new Error())
+    }
+
+    if (!request.req.headers['auth']) {
+      return done(new Error('Missing token header'))
+    }
+
+    return util
+      .promisify(jwt.verify)(request.req.headers['auth'])
+      .then(function (decoded) {
+        if (!decoded.user || !decoded.password) {
+          throw new Error('Token not valid')
+        }
+        util
+          .promisify(level.get).bind(level)(decoded.user)
+          .then(function (password) {
+            if (!password || password !== decoded.password) {
+              throw new Error('Token not valid')
+            }
+          })
+          .catch(function (err) {
+            if (err.notFound) {
+              throw new Error('Token not valid')
+            }
+            throw err
+          })
```

Comment:
I think you can condense this into the external .catch block.

Reply:
Can you, please, elaborate? For now I don't see how can this be simplified